### DISCUSSION TITLE
Deployment Issues: Undefined index: id appearing more frequently

### DISCUSSION THREAD

**Discussion by Braunson** [livewire/livewire#3954] [1. Help]
## Issue

This is caught in your error monitoring/log:

```php
Undefined index: id
```

## Cause

A user has loaded outdated assets from before a deployment, then try's to trigger an event by submitting a form or triggering a method in Livewire that makes an API call to the Livewire backend which promptly throws the noted error due to assets being out of date.

## Use Case

> A user visits a website sign up form, they take a lunch, and in the time they are away for lunch the developers of the site deploy a new release which includes updated Livewire JS assets. The user returns and try's submitting the sign up form which throws an error. 
> 
> The cause is Livewire (for example) is now expecting `updates.*.payload.id` which the outdated frontend JS the user loaded before lunch and before the deploy, are not sending and don't know to send. The solution is for the user to refresh the page and reload the newly deployed JS assets.

## My Use Case

This error has appeared a few hundred times in the past few days after deploying a new release that included an updated Livewire ([with this code](https://github.com/livewire/livewire/commit/854416a07aac67992c8a8590433b0023a896d561#)). This is critical since the site the errors have appeared on uses Livewire 100% for it's sign up / payment and login functionality. 

## Potential Solutions

There needs to be some detection of older/outdated assets in some form.

1. Add detection for `Undefined index` into the try/catch inside the `hydrate` method within `src/HydrationMiddleware/PerformActionCalls.php`. When `Undefined index` is detected within the catch block, trigger an event that persists to the frontend as a Livewire event that a developer can hook into and (for example) open a modal with some information, force a refresh or something along those lines.

2. Step #1 with the addition of returning an event, include the new version of the latest assets and force an in page reload of the Livewire assets on the fly. This way the user doesn't have to do anything and things will work as expected regardless if the assets are out of date or not. (Note: I'm not sure how feasible this is but just an idea).

## Related 

- [ErrorException: Undefined array key "id" - 🚨 out of date assets](https://github.com/livewire/livewire/discussions/3769) (Pinned Discussion for anyone who wants to troubleshoot)

**Comment by calebporzio** [comment]
Thanks for the writeup.

Just so we're clear: this situation is resolved when a user reloads the page right? Livewire's JS assets already have a cache busting mechanism to my knowledge that would clear any assets cached between pages.

This problem has reared it's head before and will certainly again, especially if we embrace the Turbo/SPA-like paradaigm.

Your second solution isn't really feasible to my knowledge unfortunately (hot-reloading with fresh assets).

Let's explore our options:
A) Provide a front-end hook called that detects stale assets and gives developers control over what to do
  - The pro with this is it's not heavy handed, the con is that many won't know to do anything with it until they're bitten
B) Automatically refresh a stale page
  - Pro: removes these errors. Pro: requires no dev customization. Con: is heavy handed and would disrupt a user on the page potentially filling out a form or something
 C) Automatically prompt the user in their browser with `confirm()`, telling them the page is out of date and asking them if they would like to refresh (just like how we handle session expiring errors)
  - Same pros as B, and less of the con of heavy handedness (but still not invisible for the user experience)

I think C is our best bet here. Before we talk implementation, what do you think @Braunson and @joshhanley?

**Reply by Braunson** [reply]
Presumably the issue is resolved when the user reloads the page but we haven't received any confirmation from the clients (no support tickets specifically about that error (nor would they) but our support would have told them to reload the page/clear cache).

I like option A as it gives flexibility to the developer to do as they want in their specific use-case rather then forcing a page reload or prompt. It could however be a hybrid between A and C, a config option, if "fire stale assets event" is set to false (could default to true), then it would fallback to option C. 

Using a hybrid of A & C we could default to C for new developers to Livewire rather then just throw an PHP error, and for anyone who wants to customize things further, they can disable the confirmation option and fire an event instead.

**Reply by joshhanley** [reply]
Yeah I agree with Braunson, a combination of A and C would be best, with the default being C, that way devs can customise it if they like.

I think the session expiry should also be setup in the same way, as I know there have been issues and PR's in the past (#2589) to try and add translation support for the session expiry message, where as a hybrid would allow them to customise as they please by using the hook.

**Reply by calebporzio** [reply]
Yeah, good point on the translation thing.

Then problem with defaulting to C, is many updates to Livewire don't actually break mid-session pages. So if we release frequent updates, we're bugging app users more often.

I think we can at least add a hook for A, and also a hook for the session expiry situation for people who want that translated

**Reply by calebporzio** [reply]
Then the next question of course is how we track different deploys without adding an entire extra param to the request payload. Maybe we put it as the first number in the checksum or something

**Reply by joshhanley** [reply]
@calebporzio yeah that is the real question! How do you track that there has been code changes? Do devs have to add a version number to their config or component or something?

I think if we answer that question, that could also solve the situation of if we default to "C" or not.

Like if it's a manual version number, then a dev can choose when to force a version, then the frontend will alert based on that, so "C" will work fine in that instance.

If it's more automatic, that'd be a different story. But I'm not sure how that'd look in php. Like do you hash the component class into a manifest to see if it's changed or something?

With JS assets you'd have a build step that generates the hash, but we don't have a build step. 

Maybe the component discovery process can add something to it's manifest for a file hash or something and fire an event if there is a hash already there but it's different?

**Reply by Braunson** [reply]
Curious if there was any progress made on this or if it's still in discussion?

Also couldn't we use this to check the version of the Livewire package

As a version number (ex. `2.10.1.0`)
```php
\Composer\InstalledVersions::getVersion('livewire/livewire');
```

As a hash, (ex. `0d417b27791af09c79108eafd1344842f83a26ee`)
```php
\Composer\InstalledVersions::getReference('livewire/livewire');
```

**Comment by joshhanley** [comment]
@Braunson yeah we recently merged this PR #4430 in which allows us to trigger an invalidation when we change things that will cause issues like the undefined key problem.

We also have posted a page to the docs with details https://laravel-livewire.com/docs/2.x/deployment

Hope this helps!
